R/contribution_matrix.R

```diff
@@ -0,0 +1,588 @@
+      mini_Lambda <- matrix(var_d/2,
+                            nrow = n_arms[studies[i]],
+                            ncol = n_arms[studies[i]])
+      #Put var_d on the leading diagonal
+      diag(mini_Lambda) <- var_d
+      #Overwrite the submatrix in Lambda_tau corresponding to this study
+      Lambda_tau[which(data_no_control$Study == studies[i]), which(data_no_control$Study == studies[i])] <- mini_Lambda
+    }
+  }
+  return(Lambda_tau)
+}
+
+
+
+#' Create the matrix of variances for exchangeable covariate parameters.
+#' 
+#' @param treatments Vector of treatments with the reference treatment first.
+#' @param std_dev_beta Standard deviation of covariate parameters.
+#' @return Lambda_beta matrix.
+CreateLambdaBetaMatrix <- function(treatments, std_dev_beta){
+  #Between study-variance
+  var_beta <- std_dev_beta^2
+  #Number of treatments
+  n_treatments <- length(treatments)
+  #Create the Lambda_tau matrix
+  Lambda_beta <- diag(var_beta, nrow = n_treatments - 1)
+  rownames(Lambda_beta) <- paste0(treatments[1], ":", treatments[-1])
+  colnames(Lambda_beta) <- paste0(treatments[1], ":", treatments[-1])
+  return(Lambda_beta)
+}
+
+
+
+#' Create the contribution matrix
+#' 
+#' @param data Input data in long format.
+#' @param treatment_ids Data frame containing treatment IDs and names in columns named 'Number' and 'Label' respectively.
+#' @param outcome_type "Continuous" or "Binary".
+#' @param effects_type "Fixed" or "Random".
+#' @param std_dev_d Between-study standard deviation. Only required when @param effects_type == "Random". Defaults to NULL.
+#' @param cov_parameters "Shared", "Exchangeable", or "Unrelated".
+#' @param std_dev_beta Standard deviation of covariate parameters. Only required when @param cov_parameters == "Exchangeable". Defaults to NULL.
+#' @param study_level TRUE for study-level contributions, FALSE for basic-comparison-level contributions.
+#' @param percentages TRUE for percentage contributions, FALSE for absolute contributions.
+#' @param all_parameters TRUE for one column per parameter, FALSE for one column per basic parameter. Defaults to TRUE.
+#' @param full_output TRUE or FALSE, defaults to FALSE. See @return for details.
+#' @return If @param full_output = FALSE:
+#'           The contribution matrix.
+#'         If @param full_output = TRUE:
+#'           List
+#'            - 'contribution' = contribution matrix.
+#'            - 'V' = The treatment effect variance matrix.
+#'            - 'X' = The design matrix.
+#'            - 'Z' = The Z matrix.
+#'            - 'Lambda_tau' = The Lambda_tau matrix (only included if @param effects_type == "Random").
+#'            - 'Lambda_beta' = The Lambda_beta matrix (only included if @param cov_parameters == "Exchangeable").
+CreateContributionMatrix <- function(data, treatment_ids, outcome_type, effects_type, std_dev_d = NULL, cov_parameters, std_dev_beta = NULL, study_level, percentages, all_parameters = TRUE, full_output = FALSE){
```

Comment:
This function is ~250 lines long, so it would be good to break it up. I'm not sure of the best division, but there seem to be smaller chunks which could be factored out.

Reply:
Created four smaller functions to be called from within this function.